### ISSUE TITLE
Flaky ListenersOnReconnectTest

### ISSUE THREAD

**Issue by puzpuzpuz** [hazelcast/hazelcast-nodejs-client#666]
http://jenkins.hazelcast.com/job/NodeJS-10-windows-4.0.x/16/testReport/junit/kill%20two%20members%20[0,2],%20listener%20still%20receives%20map/put%20event%20[smart=false]/ListenersOnReconnectTest_kill_two_members__0_2___listener_still_receives_map_put_event__smart_false_/

```
Error Message
Heartbeat timed out
Stacktrace
Error: Heartbeat timed out
    at new TargetDisconnectedError (lib\core\HazelcastError.js:18:10274)
    at Timeout.Util_1.scheduleWithRepetition [as _onTimeout] (lib\invocation\InvocationService.js:35:2688)
```

**Comment by aslatter** [comment]
This looks similar to an issue I just ran into - if I bounce Hazelcast, my Hazelcase-nodejs-client which is already connected waits ~1 minute to reconnect when we next try to use the connection, instead of reconnecting much more quickly.

In my debugging, the issue is that the socket-close handler isn't able to pro-actively clean up the connection.

ConnectionManager has this attempt at cleanup - when we set up the initial TLS connection we add an error-callback to the Node.js socket to proactively clean-up broken sockets:

```ts
        socket.on('error', (e: any) => {
            this.logger.warn('ConnectionManager', 'Could not connect to address ' + address.toString(), e);
            connectionResolver.reject(e);
            if (e.code === 'EPIPE' || e.code === 'ECONNRESET') {
                const connection = this.getConnectionFromAddress(address);
                if (connection != null) {
                    this.onConnectionClose(connection);
                }
            }
        });
```

However *getConnectionFromAddress* fails, because the 'address' we are using for lookup doesn't match the 'remoteAddress' stored in the 'ClientConnection' object.

I don't quite know why this is, but I suspect it's because we update the ClientConnection 'remoteAddress' with the response from authentication:

authenticateOnCluster -> handleSuccessfulAuth -> `connection.setRemoteAddress(response.address);`

In my case, the 'address' we are using for lookup is "localhost:port", but the 'remoteAddress' we fail to match to is "<my IP address>:port".

Let me know if you'd rather have this in a separate issue.

**Comment by aslatter** [comment]
Ah - my example fails on Linux and Windows, if that's helful. I just noticed this issue has "Windows" in its name.

**Comment by puzpuzpuz** [comment]
@aslatter thanks for reporting this issue. It indeed sounds like something that is worth a separate GH issue, so I'd appreciate if you could create a separate one.

As for the reproducer, what do you do to make it happen? Just restart a member in 2+ member cluster or it's something more tricky?

**Comment by puzpuzpuz** [comment]
@aslatter I've created #704 and there is a good chance that it will fix the issue. It would be awesome if you could try that patch.

**Comment by puzpuzpuz** [comment]
Saw this one on `master` branch and Linux today: http://jenkins.hazelcast.com/job/NodeJS-pr-builder/804/testReport/junit/restart%20member,%20listener%20still%20receives%20map/put%20event%20[smart=true]/ListenersOnReconnectTest_restart_member__listener_still_receives_map_put_event__smart_true_/

**Comment by aslatter** [comment]
That patch looks significantly better in my test-case.

What I was doing was basically the same as the automated test you've added in that patch - I was using the restart of a single-node-cluster as a proxy for testing how the client behaves during "cluster events".

Thanks for the quick response!

**Comment by puzpuzpuz** [comment]
@aslatter thanks for the confirmation and thanks once again for reporting the issue. I'll make sure to include the patch into 4.0.2 release.

**Comment by puzpuzpuz** [comment]
Closed with #718

**Comment by puzpuzpuz** [comment]
Reopening as it occurred during a recent build on Windows + Node.js v10:
```
kill two members [0,2], listener still receives map.put event [smart=true].ListenersOnReconnectTest kill two members [0,2], listener still receives map.put event [smart=true] (from ListenersOnReconnectTest)

Error Message
null
Stacktrace
Error: null
    at new TargetDisconnectedError (lib\core\HazelcastError.js:18:10399)
    at Timeout.Util_1.scheduleWithRepetition [as _onTimeout] (lib\invocation\InvocationService.js:35:2738)
```

http://jenkins.hazelcast.com/job/NodeJS-10-windows-master/230/testReport/junit/kill%20two%20members%20[0,2],%20listener%20still%20receives%20map/put%20event%20[smart=true]/ListenersOnReconnectTest_kill_two_members__0_2___listener_still_receives_map_put_event__smart_true_/

**Comment by puzpuzpuz** [comment]
Should be closed with #728

**Comment by srknzl** [comment]
Reopening due to a fail in gh actions in backward compatibility test of 4.0.0 client

Note that the change in #728 were applied in the running test.

https://github.com/hazelcast/client-compatibility-test/runs/2390350977?check_suite_focus=true

We might want to increase the sleep amount more.
